CredentialsManager/App.cs

```diff
@@ -0,0 +1,161 @@
+        try
+        {
+          Manager.Initialize();
+          PrintMenu();
+          GetUserInput();
+        }
+        catch (MyFileNotFoundException ex)
+        {
+          Console.WriteLine(ex.Message);
+          break;
+        }
+        catch (DuplicateUserCredentialsException ex)
+        {
+          Console.WriteLine(ex.Message);
+          Pause("Press ENTER to exit..");
+          break;
+        }
+        catch (UserNameIsTakenException ex)
+        {
+          Console.WriteLine(ex.Message);
+          Pause();
+        }
+      }
+    }
+
+    private static IList<MenuItem> GetMenuItems()
+    {
+      return new List<MenuItem>
+      {
+        new MenuItem { ItemId = 1, Label = "Login", Action = Login, ConsoleKey = ConsoleKey.D1 },
+        new MenuItem { ItemId = 2, Label = "Register", Action = Register, ConsoleKey = ConsoleKey.D2 }
+      };
+    }
+
+    private static Menu CreateMenu()
+    {
+      return new Menu
+      {
+        Header = "Credentials Manager",
+        MenuItems = GetMenuItems(),
+        ExitItem = new MenuItem { ItemId = 3, Label = "Exit", Action = Exit, ConsoleKey = ConsoleKey.D3 }
+      };
+    }
+
+    private static void ConsoleInit(string header)
+    {
+      Console.CursorVisible = false;
+      Console.Clear();
+      Console.WriteLine(header);
+      Console.WriteLine();
+    }
+
+    private static void PrintMenu()
+    {
+      ConsoleInit(Header);
+
+      Console.WriteLine(MainMenu.ToString());
+    }
+
+    private static void GetUserInput()
+    {
+      ConsoleKey pressedKey;
+      do
+      {
+        pressedKey = Console.ReadKey(true).Key;
+        if (!MainMenu.IsValidKey(pressedKey)) continue;
+
+        ConsoleInit(MainMenu[pressedKey].Label);
+        MainMenu[pressedKey].Action();
+        Pause();
+        break;
+      } while (pressedKey != MainMenu.ExitItem.ConsoleKey);
+    }
+
+    private static void Login()
+    {
+      var userName = PromptString("Username: ");
+      var userPassword = PromptString("Password: ", false);
+
+      var isLoginSuccessful = Manager.Login(new Credentials(userName, userPassword));
+      var statusMessage = isLoginSuccessful ? LoginMessage : FailedLoginMessage;
+
+      Console.WriteLine(statusMessage);
+    }
+
+    private static void Register()
+    {
+      var userName = PromptString("Username: ");
+      var userPassword = PromptString("Password: ", false);
+
+      Manager.Register(new Credentials(userName, userPassword));
+      Console.WriteLine(RegistrationSuccessfulMessage);
+    }
+
+    private static void Exit()
+    {
+      _isExitInvoked = true;
+    }
+
+    private static string PromptString(string message)
+    {
+      string userInput;
+      do
+      {
+        Console.Write(message);
+        userInput = Console.ReadLine() ?? string.Empty;
+      } while (string.IsNullOrWhiteSpace(userInput));
+
+      return userInput;
+    }
+
+    private static string PromptString(string message, bool isKeyVisible)
```

Comment:
I like the cool tricks to hide the password.  I do not like the fact that you have two methods named `PromptString` like this.  It seems `PromptString(message)` is really just `PromptString(message, true)`.  Also, I don't think you call `PromptString(message, true)` anywhere in your code.

Reply:
I decided to remove the PromptString overload.
I updated in commit [17deff9](https://github.com/csinn/CSharp-From-Zero-To-Hero-v2/pull/43/commits/17deff93537954ed330d11ab0d6b08e1324dd225)